cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/CucumberExpressionException.java

```diff
@@ -1,14 +1,115 @@
 package io.cucumber.cucumberexpressions;
 
+import io.cucumber.cucumberexpressions.Ast.AstNode;
+import io.cucumber.cucumberexpressions.Ast.Token;
+import io.cucumber.cucumberexpressions.Ast.Token.Type;
 import org.apiguardian.api.API;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
 @API(status = API.Status.STABLE)
 public class CucumberExpressionException extends RuntimeException {
+
     CucumberExpressionException(String message) {
         super(message);
     }
 
     CucumberExpressionException(String message, Throwable cause) {
         super(message, cause);
     }
+
+    static CucumberExpressionException createCaptureGroupParameterTypeMisMatch(TreeRegexp treeRegexp,
+            List<ParameterType<?>> parameterTypes, List<Group> argGroups) {
+        return new CucumberExpressionException(
+                String.format("Expression /%s/ has %s capture groups (%s), but there were %s parameter types (%s)",
+                        treeRegexp.pattern().pattern(),
+                        argGroups.size(),
+                        getGroupValues(argGroups),
+                        parameterTypes.size(),
+                        getParameterTypeNames(parameterTypes)
+                ));
+    }
+
+    private static List<String> getParameterTypeNames(List<ParameterType<?>> parameterTypes) {
+        List<String> list = new ArrayList<>();
+        for (ParameterType<?> type : parameterTypes) {
+            String name = type.getName();
+            list.add(name);
+        }
+        return list;
+    }
+
+    private static List<Object> getGroupValues(List<Group> argGroups) {
+        List<Object> list = new ArrayList<>();
+        for (Group argGroup : argGroups) {
+            String value = argGroup.getValue();
+            list.add(value);
+        }
+        return list;
+    }
+
+    static CucumberExpressionException createMissingEndTokenException(Type beginToken, Type endToken,
+            List<Token> tokens, int current) {
+        int index = currentTokenIndex(tokens, current);
+        return new CucumberExpressionException("" +
```

Comment:
This follow the pattern of pre-amble, new line, expression, pointer, problem description, suggested solution.